### ISSUE TITLE
Cloudflare CAPTCHA

### ISSUE THREAD

**Issue by m4mb01t4l14n0** [OnionBrowser/iOS-OnionBrowser#256]
I opened duckduckgo search and, on several of the result sites I received a "block page" (attached) that asked me to complete a CAPTCHA.  All three times, this came from Cloudflare.  Page asked for the typical tap on the "I am a human" checkbox.  Unfortunately, I was not able to actually CHECK.  Tapping on the checkbox "selected" it, but no highlight mark appeared and tapping SUBMIT got the response shown.

We have an issue #145 (Integration with Cloudflare’s captcha token identity system) as stretch-goal item.  Is this it?  Somewhat amazing that upon my first search using 2.6.0 3 of the first 4 search results were hosted on CloudFlare and gave me that CAPTCHA (coincidence!). 
#145 

![IMG_0017](https://user-images.githubusercontent.com/34868071/77583942-4d693600-6eb8-11ea-9f69-fc5d2e3010e3.PNG)

**Comment by tladesignz** [comment]
I investigated #145, but never made the effort to comment on it. I did now: https://github.com/OnionBrowser/OnionBrowser/issues/145#issuecomment-604369296

TL;DR: It seems, we're unable to improve here.

**Comment by tladesignz** [comment]
Problem might also be found here: https://www.webmd.com/

**Comment by tladesignz** [comment]
Ok, so I found a viable solution to make the CAPTCHA work:

Cloudflare adds some specific headers to the response. If that is seen, Onion Browser now allows `Content-Security-Policy: connect-src: hcaptcha.com`, so the Captcha works.

However, for me, I run into a captcha loop. Not sure, if this is on all sites, from all IP addresses. Might have to do with the specific Cloudflare configuration, or it might be a bug triggered by something Onion Browser does (or does not).

Hard to debug, though, because of the immediate reload.

I guess, what would be a good idea, is to implement this to make sure, the Captcha stays away for a while:

https://privacypass.github.io/
https://support.cloudflare.com/hc/en-us/articles/115001992652-Privacy-Pass
https://github.com/privacypass/challenge-bypass-extension

**Comment by m4mb01t4l14n0** [comment]
Just discovered Cloudflare moved from Google's reCaptcha to hCaptcha in April 2020.  
https://latesthackingnews.com/2020/04/14/cloudflare-announces-plans-to-dump-recaptcha-for-hcaptcha/

**Comment by m4mb01t4l14n0** [comment]
@tladesignz what's your estimate of time to implement Privacy Pass in a non-Chrome browser like Onion Browser?

**Comment by tladesignz** [comment]
Ok, so I dug a little deeper into that privacy-pass plugin. It relies heavily on being run as a browser plugin (e.g. runs continuously in the background, makes use of hooks only available to plugins) and we can't run it just like that in a UIWebView context.
I would need to change it completely, to run injected inside a website, only.

I sent @alxdavids, the main committer on that project a mail to ask about his opinion before I waste more time on this.

I also just played around with desktop Tor Browser, and it seems to have the exact same problem: The hCaptcha is shown over and over again. Adding that plugin seems to help, though.
(Also, I have to admit, it isn't shown every time. There seems to be some credibility calculation on CloudFlare's side and Tor Browser might to be nearer to the threshold than Onion Browser. May have something to do with exit nodes, too.)

**Comment by m4mb01t4l14n0** [comment]
This is incredibly interesting, really.  The challenge for me is that this problem does not happen just on a few rinky-dink websites.  It happens on many of the sites people really want to use.  It's a terrible impediment, making Onion Browser itself appear to be the problem.    I'll be interested to hear what the plug-in's author says.

**Comment by zettysource** [comment]
So there is no way to use on iOS?

**Comment by mtigas** [comment]
@zettysource As I mentioned on another thread you commented in, you can download Onion Browser for iOS in the App Store: https://apps.apple.com/us/app/onion-browser/id519296448

You are replying to different unrelated e-mail threads, which are for to-do tasks and bug reports of the Onion Browser app. Please do not do that. This is not a mailing list. (My previous replies to you are located in the thread here: https://github.com/OnionBrowser/OnionBrowser/pull/273 )

**Comment by zettysource** [comment]
> I opened duckduckgo search and, on several of the result sites I received a "block page" (attached) that asked me to complete a CAPTCHA. All three times, this came from Cloudflare. Page asked for the typical tap on the "I am a human" checkbox. Unfortunately, I was not able to actually CHECK. Tapping on the checkbox "selected" it, but no highlight mark appeared and tapping SUBMIT got the response shown.
> 
> We have an issue #145 (Integration with Cloudflare’s captcha token identity system) as stretch-goal item. Is this it? Somewhat amazing that upon my first search using 2.6.0 3 of the first 4 search results were hosted on CloudFlare and gave me that CAPTCHA (coincidence!).
> #145
> 
> ![IMG_0017](https://user-images.githubusercontent.com/34868071/77583942-4d693600-6eb8-11ea-9f69-fc5d2e3010e3.PNG)

**Comment by zettysource** [comment]
#273

**Comment by tladesignz** [comment]
Ok, so I got an answer from Alex.
The thing is: that plugin intercepts and manipulates HTTP traffic, which we can't do with JavaScript inside a webpage, because that only runs after the fact.

We of course do have access to and are able to manipulate HTTP headers, but from outside, meaning with Swift/Objective-C.
To try to rewrite that part of the code in Swift/Objective-C and then keep an instance of `UIWebView`/`WKWebView` around where the JavaScript plugin code is doing all its crypto seems like a non-effortless and quite strange setup.

In other words: Doing it right, means, replicating the complete plugin in Swift, doing it less right still seems to be an awful lot of work. Nothing I could do in the remaining days of the grant.

So, this needs to be postponed, sorry.

**Comment by m4mb01t4l14n0** [comment]
@tladesignz let's close this issue.  I think the verbal documentation here is sufficient to indicate we can not, under current funding and time constraints, act further on this item.

**Comment by m4mb01t4l14n0** [comment]
my only comment is "<sad face emoji>"
